src/Build/Graph/ProjectGraph.cs

```diff
@@ -243,6 +297,92 @@ private static ImmutableList<string> ExpandDefaultTargets(ProjectInstance projec
             return targets;
         }
 
+        /// <summary>
+        /// Gets the effective global properties for a project reference.
+        /// </summary>
+        /// <remarks>
+        /// The behavior of this method should match the logic in Microsoft.Common.CurrentVersion.targets and the MSBuild task.
+        /// </remarks>
+        private static PropertyDictionary<ProjectPropertyInstance> GetProjectReferenceGlobalProperties(ProjectItemInstance projectReference, PropertyDictionary<ProjectPropertyInstance> requesterGlobalProperties)
+        {
+            string propertiesString = projectReference.GetMetadataValue(PropertiesMetadataName);
+            string additionalPropertiesString = projectReference.GetMetadataValue(AdditionalPropertiesMetadataName);
+            string undefinePropertiesString = projectReference.GetMetadataValue(UndefinePropertiesMetadataName);
+            string globalPropertiesToRemoveString = projectReference.GetMetadataValue(GlobalPropertiesToRemoveMetadataName);
+
+            // TODO: Mimic AssignProjectConfiguration's behavior for determining the values for these.
+            string setConfigurationString = projectReference.GetMetadataValue(SetConfigurationMetadataName);
+            string setPlatformString = projectReference.GetMetadataValue(SetPlatformMetadataName);
+            string setTargetFrameworkString = projectReference.GetMetadataValue(SetTargetFrameworkMetadataName);
+
+            // The properties on the project reference supercede the ones from the MSBuild task instad of appending.
```

Comment:
Also: really? That is confusing and sad.

Reply:
Unless I'm reading the MSBuild task incorrectly, yea :(. I suspect that's why `AdditionalProperties` even exists. `UndefineProperties` seems to behave in the expected way.